### PR TITLE
implement Secret Karma configuration

### PR DESCRIPTION
**Pull request by dminca**
if the Karma config contains sensitive information you wouldn't want to
accidendatlly commit this into your git repo, therefore we need a way
to provide safe access to the configuration but also allow storing this
configuration encrypted in git (by any means possible).

Kubernetes Secrets alone are not secure, but they can be improved
through additional apps (ie. SealedSecret Operator etc.), therefore, by
storing this sensitive configuration in a Secret (encrypted) not only
we're avoiding bad practice of pushing sensitive information in
plain-text to git repos, but also making further room for improving the
Security & Compliance of the chart.

Resolves: #20
Related:
Signed-off-by: Daniel-Andrei Minca <mandrei17@gmail.com>

### CODE REVIEW

**Review by drazul — APPROVED** [review]

File: charts/karma/README.md [hunk 1]
```diff
@@ -52,3 +52,51 @@ $ helm install --name my-release -f values.yaml stable/karma
...
+# create a Secret from this file with the key `karma.conf`
+kubectl create secret generic sensitive-karma-config \
+    --from-file=karma.conf=./karma.conf
+```
+
+Next, point to the secret in your values file
+
+```yml
+# values.yaml
+secretConfig:
```
Comment (desaintmartin) [inline charts/karma/README.md:97]: What about `existingSecretConfig`?

File: charts/karma/templates/deployment.yaml [hunk 2]
```diff
@@ -93,11 +93,16 @@ spec:
 {{ toYaml . | indent 8 }}
     {{- end }}
       volumes:
-      {{- if .Values.configMap.enabled }}
+      {{- if and (.Values.configMap.enabled) (not .Values.secretConfig.enabled) }}
       - name: karma-config
         configMap:
           name: {{ .Release.Name }}-config
       {{- end }}
+      {{- if and (.Values.secretConfig.enabled) (not .Values.configMap.enabled) }}
```
Comment (desaintmartin) [inline charts/karma/templates/deployment.yaml:101]: You could forbid to have the two parameters enabled:
```yaml
{{- if and .Values.secretConfig.enabled .Values.configMap.enabled }}
{{- fail "plz don't" }}
{{- end }}
```

And then simplify those conditions.

**Review by desaintmartin — CHANGES_REQUESTED** [review]
Thanks for the contribution and sorry for the delay!
Note that we could also change the ConfigMap to be a Secret instead, by default, which could allow to have secured data inside of kube IF the values.yaml is securely generated.

File: charts/karma/README.md (hunk 1 above)
Comment (dminca) [reply charts/karma/README.md]: good point, applied it

File: charts/karma/templates/deployment.yaml (hunk 2 above)
Comment (dminca) [reply charts/karma/templates/deployment.yaml]: That's a really good suggestion, thx for this one, I never used `fail` in Helm templates so far :)

**Comment by dminca** [comment]
> Thanks for the contribution and sorry for the delay!
> Note that (in another PR) we could also change the ConfigMap to be a Secret instead, by default, which could allow to have secured data inside of kube IF the values.yaml is securely generated.

This one's a little bit tricky because we'll enforce the same behaviour as with the `ConfigMap` which is what I wanted to avoid initially -- people storing _sensitive_ data in their repos during chart provisioning.

By simply pointing to a pre-created `Secret` by the user, they'll not push this in the chart provisioning `values.yaml`, what do you think?

Lastly, by having a separate `Secret` provisioned, they can potentially encrypt this via the `SealedSecret` Operator and push it to the repo (as it's safe anyway)

File: charts/karma/templates/deployment.yaml (hunk 2 above)
Comment (desaintmartin) [reply charts/karma/templates/deployment.yaml]: Great!

**Comment by desaintmartin** [comment]
Oh, sure, it was more a note to a future self. We generate the values.yaml in CI/CD, so it makes sense for us, but those are in fact two different and valid use cases.

Sorry, I've just upadted karma, could you rebase?

**Comment by dminca** [comment]
> Oh, sure, it was more a note to a future self. We generate the values.yaml in CI/CD, so it makes sense for us, but those are in fact two different and valid use cases.
> 
> Sorry, I've just upadted karma, could you rebase?

Yup, rebased :)

**Review by desaintmartin — APPROVED** [review]

File: charts/karma/templates/deployment.yaml [hunk 3]
```diff
@@ -93,11 +93,19 @@ spec:
 {{ toYaml . | indent 8 }}
     {{- end }}
       volumes:
+      {{- if .Values.configMap.enabled .Values.existingSecretConfig.enabled }}
```
Comment (desaintmartin) [inline charts/karma/templates/deployment.yaml:96]:
```suggestion
      {{- if and .Values.configMap.enabled .Values.existingSecretConfig.enabled }}
```

Comment (desaintmartin) [reply charts/karma/templates/deployment.yaml:96]: Sorry, did not see that earlier

Comment (dminca) [reply charts/karma/templates/deployment.yaml]: ah, nicely spotted, thank you 😄

**Comment by desaintmartin** [comment]
Thanks!

**Pull request merged**